src/tls/openssl.c

```diff
@@ -724,62 +752,64 @@ static int tls_net_handshake(struct flb_tls *tls,
         }
     }
 
+    // Clear any previous SSL errors
     ERR_clear_error();
 
+    // Perform SSL handshake
     if (tls->mode == FLB_TLS_CLIENT_MODE) {
         ret = SSL_connect(session->ssl);
     }
     else if (tls->mode == FLB_TLS_SERVER_MODE) {
         ret = SSL_accept(session->ssl);
     }
 
-    if (ret != 1) {
-        ret = SSL_get_error(session->ssl, ret);
-        if (ret != SSL_ERROR_WANT_READ &&
-            ret != SSL_ERROR_WANT_WRITE) {
-            ret = SSL_get_error(session->ssl, ret);
-            // The SSL_ERROR_SYSCALL with errno value of 0 indicates unexpected
-            //  EOF from the peer. This is fixed in OpenSSL 3.0.
-            if (ret == 0) {
+    // Exit early if the handshake was successful
+    if (ret == 1) {
+        session->continuation_flag = FLB_FALSE;
+        pthread_mutex_unlock(&ctx->mutex);
+        flb_debug("[tls] connection and handshake ok");
+        return 0;
+    }
+
+    int ssl_err = SSL_get_error(session->ssl, ret);
+
+    // Handle SSL handshake errors
+    switch (ssl_err) {
+        case SSL_ERROR_WANT_READ:
+            // Handle non-blocking read operation
+            pthread_mutex_unlock(&ctx->mutex);
+            session->continuation_flag = FLB_TRUE;
+            return FLB_TLS_WANT_READ;
+
+        case SSL_ERROR_WANT_WRITE:
+            // Handle non-blocking write operation
+            pthread_mutex_unlock(&ctx->mutex);
+            session->continuation_flag = FLB_TRUE;
+            return FLB_TLS_WANT_WRITE;
+
+        case SSL_ERROR_SYSCALL:
+        case SSL_ERROR_SSL:
+        default:
+            // Handle unexpected EOFs and other errors
+            if (ssl_err == 0) {
                 ssl_code = SSL_get_verify_result(session->ssl);
                 if (ssl_code != X509_V_OK) {
-                    flb_error("[tls] error: unexpected EOF with reason: %s",
-                              ERR_reason_error_string(ERR_get_error()));
+                    const char *x509_err = X509_verify_cert_error_string(ssl_code);
```

Comment:
Probably, we encourage to define this variable on the top part of this function.
https://github.com/fluent/fluent-bit/blob/master/CONTRIBUTING.md#variable-definitions